README.md
# Exbet-API bindings for Python 3

![exbetapi](https://github.com/exbet/python-exbetapi/workflows/tox/badge.svg?branch=develop)
![exbetapi](https://github.com/exbet/python-exbetapi/workflows/tox/badge.svg?branch=master)
[![Documentation Status](https://readthedocs.org/projects/python-exbetapi/badge/?version=master)](https://python-exbetapi.readthedocs.io/en/master/?badge=master)


## Installation

    pip3 install exbetapi

## Usage

After installation, you can either use exbetapi as a library or use the
command-line interface tool (CLI) that comes with it.

### CLI tool

Show the CLI help:

    $ eb --help
    $ eb <command> --help

To show user details, use:

    $ eb account
    User: <username>
    Password: <password>

To list your bets

    $ eb list-bets

To place a new bet

    $ eb placebet SELECTION_ID [back|lay] BACKER_MULTIPLIER BACKER_STAKE
    # for instance
    $ eb placebet 1.25.38235 back 2.5 0.00001

### Library

To use the library, you first need to instanciate `ExbetAPI` and login:

```python
from exbetapi import ExbetAPI
api = ExbetAPI()
api.login(user, password)
```

After that you can get your account details with:

```python
print(api.account)
```

List available sports, groups, markets and selections:

```python
print(api.list_sports())
print(api.lookup_eventgroups(
    sport_id
))
print(api.list_events(
    eventgroup_id
))
print(api.list_markets(
    event_id
))
print(api.list_selections(
    market_id
))
```

List your bets with:

```python
print(api.list_bets())
```

... or place a new bet using:

```python
api.place_bet(
    selection_id,
    back_or_lay,
    backer_multiplier,
    "{} BTC".format(backer_stake),
    bool(persistent),
)
```

Full Documentation is available [here](https://python-exbetapi.readthedocs.io/en/master/exbetapi.api.html). The underlying server-side
APIs can be found [here](https://api.macau.exbet.io/apidocs/).

## Development use

Clone the repository and create and activate a python virtual enviroment (assumes virtualenv is installed)

    git clone https://github.com/exbet/python-exbetapi.git
    cd python-exbetapi
    virtualenv env --python python3
    source env/bin/activate

Install Poetry and Exbet API dependencies

    pip3 install poetry
    poetry install

Run tests

    poetry run pytest

## Documentation

Documentation is available [here](https://python-exbetapi.rtfd.io)!


CHANGELOG.md
# Changelog
Note: version releases in the 0.x.y range may introduce breaking changes.

## 2.1.1

- patch: Allow to change the BASE URL for testing
- patch: Allow to specify the new base url via cli
- patch: Improve README.md
- patch: Install 'eb' as cli tool

## 2.1.0

- minor: connect to macau deployment per default
- patch: Add a User-agent to the requests

## 2.0.0

- major: api renaming (breaking)

## 1.0.0

- major: api renaming (breaking)

## 0.1.0

- minor: adapt to api changes



tests/test_exbetapi.py
# -*- coding: utf-8 -*-
import unittest
import pytest
from exbetapi import ExbetAPI
from exbetapi.exceptions import AlreadyLoggedinException
from functools import wraps


def API_POST(d):
    return unittest.mock.patch("exbetapi.ExbetAPI._post", return_value=d)


def API_GET(d):
    return unittest.mock.patch("exbetapi.ExbetAPI._get", return_value=d)


def API_PROPERTY(prop, d):
    return unittest.mock.patch(
        "exbetapi.ExbetAPI." + prop,
        new_callable=unittest.mock.PropertyMock,
        return_value=d,
    )


@pytest.fixture(scope="session")
@API_POST(dict(access_token="a", refresh_token="b", username="u"))
def api(mockery):
    """ This api instance will be logged in (fake)
    """
    a = ExbetAPI()
    a.login("", "")
    mockery.assert_called_once_with("login", dict(username="", password=""))
    return a


def test_init():
    """ Ensure we can instantiate the API
    """
    ExbetAPI()


def test_no_headers():
    assert "Authorization" not in ExbetAPI()._headers()


def test_headers(api):
    assert api._headers().get("Authorization") == "Bearer a"


@unittest.mock.patch("requests.post")
def test_post(m, api):
    from exbetapi import __version__

    api.get_bet("1.26.0")
    m.assert_called_once_with(
        ExbetAPI.BASEURL + "bet/get",
        json={"bet_id": "1.26.0"},
        headers={
            "Authorization": "Bearer a",
            "User-Agent": "python/exbetapi-" + __version__,
        },
    )


@unittest.mock.patch("requests.get")
def test_get(m, api):
    api.account


def test_login(api):
    """ Test that we are logged in
    """
    assert api.is_loggedin()


def test_parse_stake(api):
    assert api._parse_stake("1 BTC") == dict(amount=1, symbol="BTC")
    assert api._parse_stake(dict(amount=1, symbol="BTC")) == dict(
        amount=1, symbol="BTC"
    )


def test_relogin(api):
    with pytest.raises(AlreadyLoggedinException):
        api.login("", "")


def test_reset(api):
    assert api.is_loggedin()
    api.reset()
    assert not api.is_loggedin()


@API_GET(dict(foo="bar"))
def test_account(mock, api):
    assert api.account == dict(foo="bar")
    mock.assert_called_once_with("account")


@API_GET(dict(foo="bar"))
def test_info(mock, api):
    assert api.info == dict(foo="bar")
    mock.assert_called_once_with("info")


@API_GET(dict(balances=[dict(amount=10, symbol="BTC")]))
def test_balance(mock, api):
    assert api.balance == dict(BTC=10)
    mock.assert_called_once_with("balance")


@API_GET(dict(foo="bar"))
def test_session(mock, api):
    assert api.session == dict(foo="bar")
    mock.assert_called_once_with("session")


@API_GET(dict(roles=["foobar"]))
def test_roles(mock, api):
    assert api.roles == ["foobar"]
    mock.assert_called_once_with("account")


@API_GET(dict(matched=[], unmatched=[]))
def test_list_bets(mock, api):
    assert api.list_bets() == dict(matched=[], unmatched=[])
    mock.assert_called_once_with("bet/list")


@API_POST(dict(sports=["foobar"]))
def test_lookup_sports(mock, api):
    assert api.lookup_sports() == ["foobar"]
    mock.assert_called_once_with("lookup/sports", dict())


@API_POST(dict(eventgroups=["foobar"]))
def test_lookup_eventgroups(mock, api):
    _id = "1.20.0"
    assert api.lookup_eventgroups(_id) == ["foobar"]
    mock.assert_called_once_with("lookup/eventgroups", dict(sport_id=_id))


@API_POST(dict(events=["foobar"]))
def test_lookup_events(mock, api):
    _id = "1.21.0"
    assert api.lookup_events(_id) == ["foobar"]
    mock.assert_called_once_with("lookup/events", dict(eventgroup_id=_id))


@API_POST(dict(foo="bar"))
def test_lookup_event(mock, api):
    _id = "1.22.0"
    assert api.lookup_event(_id) == dict(foo="bar")
    mock.assert_called_once_with("lookup/event", dict(event_id=_id))


@API_POST(dict(markets=["foobar"]))
def test_lookup_markets(mock, api):
    _id = "1.22.0"
    assert api.lookup_markets(_id) == ["foobar"]
    mock.assert_called_once_with("lookup/markets", dict(event_id=_id))


@API_POST(dict(foo="bar"))
def test_lookup_market(mock, api):
    _id = "1.24.0"
    assert api.lookup_market(_id) == dict(foo="bar")
    mock.assert_called_once_with("lookup/market", dict(market_id=_id))


@API_POST(dict(selections=["foobar"]))
def test_lookup_selections(mock, api):
    _id = "1.24.0"
    assert api.lookup_selections(_id) == ["foobar"]
    mock.assert_called_once_with("lookup/selections", dict(market_id=_id))


@API_POST(dict(foo="bar"))
def test_lookup_selection(mock, api):
    _id = "1.25.0"
    assert api.lookup_selection(_id) == dict(foo="bar")
    mock.assert_called_once_with("lookup/selection", dict(selection_id=_id))


@API_POST(dict(aggregated_back_bets=[], aggregated_lay_bets=[]))
def test_orderbook(mock, api):
    _id = "1.25.0"
    assert api.orderbook(_id) == dict(aggregated_back_bets=[], aggregated_lay_bets=[])
    mock.assert_called_once_with("lookup/orderbook", dict(selection_id=_id))


@API_POST(dict())
def test_placebet(mock, api):
    api.place_bet("1.25.50449", "lay", 1.65, "0.01 BTC")
    mock.assert_called_once_with(
        "bet/place",
        dict(
            back_or_lay="lay",
            selection_id="1.25.50449",
            backer_multiplier=1.65,
            persistent=True,
            backer_stake=dict(amount=0.01, symbol="BTC"),
        ),
    )


@API_POST(dict())
def test_placebets(mock, api):
    api.place_bets(
        [
            ("1.25.50449", "lay", 1.65, "0.01 BTC"),
            ("1.25.50449", "back", 1.75, "0.01 BTC"),
        ]
    )
    mock.assert_called_once_with(
        "bet/place_many",
        {
            "place_bets": [
                {
                    "selection_id": "1.25.50449",
                    "back_or_lay": "lay",
                    "backer_multiplier": 1.65,
                    "backer_stake": {"amount": 0.01, "symbol": "BTC"},
                    "persistent": True,
                },
                {
                    "selection_id": "1.25.50449",
                    "back_or_lay": "back",
                    "backer_multiplier": 1.75,
                    "backer_stake": {"amount": 0.01, "symbol": "BTC"},
                    "persistent": True,
                },
            ]
        },
    )


@API_POST(dict())
def test_placebet_nonpersistent(mock, api):
    api.place_bet("1.25.50449", "lay", 1.65, "0.01 BTC", False)
    mock.assert_called_once_with(
        "bet/place",
        dict(
            back_or_lay="lay",
            selection_id="1.25.50449",
            backer_multiplier=1.65,
            persistent=False,
            backer_stake=dict(amount=0.01, symbol="BTC"),
        ),
    )


@API_POST(dict())
def test_placebets_nonpersistent(mock, api):
    api.place_bets(
        [
            ("1.25.50449", "lay", 1.65, "0.01 BTC", False),
            ("1.25.50449", "back", 1.75, "0.01 BTC", False),
        ]
    )
    mock.assert_called_once_with(
        "bet/place_many",
        {
            "place_bets": [
                {
                    "selection_id": "1.25.50449",
                    "back_or_lay": "lay",
                    "backer_multiplier": 1.65,
                    "backer_stake": {"amount": 0.01, "symbol": "BTC"},
                    "persistent": False,
                },
                {
                    "selection_id": "1.25.50449",
                    "back_or_lay": "back",
                    "backer_multiplier": 1.75,
                    "backer_stake": {"amount": 0.01, "symbol": "BTC"},
                    "persistent": False,
                },
            ]
        },
    )


@API_POST(dict())
def test_cancelbet(mock, api):
    api.cancel_bet("1.26.171197")
    mock.assert_called_once_with("bet/cancel", {"bet_id": "1.26.171197"})


@API_POST(dict())
def test_cancelbets(mock, api):
    api.cancel_bets(["1.26.171196", "1.26.171195"])
    mock.assert_called_once_with(
        "bet/cancel_many",
        {"bets_to_cancel": [{"bet_id": "1.26.171196"}, {"bet_id": "1.26.171195"}]},
    )


@API_POST(dict())
def test_get_task(mock, api):
    api._get_task("asfasf")
    mock.assert_called_once_with("task", {"task_id": "asfasf"})


@API_POST(dict())
def test_find_selection(mock, api):
    api.find_selection("Sport", "Eventgroup", ["team1", "team2"], "market", "selection")
    mock.assert_called_once_with(
        "find/selections",
        {
            "sport": "Sport",
            "event_group": "Eventgroup",
            "teams": ["team1", "team2"],
            "market": "market",
            "selection": "selection",
        },
    )


if __name__ == "__main__":
    unittest.main()


exbetapi/cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" CLI Tools for exbetapi
"""

# -*- coding: utf-8 -*-

import json
import click
from exbetapi import ExbetAPI

api = None


def dump(x):
    """Dump json content"""
    click.echo(json.dumps(x, indent=4))


@click.group()
@click.option("--user", prompt="User", required=True)
@click.option("--password", prompt="Password", hide_input=True)
@click.option("--test/--no-test", default=False)
def main(user, password, test):
    """CLI tools for exbetapi"""
    global api
    if test:
        api = ExbetAPI(use_everett=True)
    else:
        api = ExbetAPI()
    api.login(user, password)


@main.command()
def account():
    """Show account details"""
    dump(api.account)


@main.command()
def info():
    """Show API infos"""
    dump(api.info)


@main.command()
def balance():
    """Show account balance"""
    dump(api.balance)


@main.command()
def session():
    """Show API session info"""
    dump(api.session)


@main.command()
def roles():
    """Show account roles"""
    dump(api.roles)


@main.command()
def list_bets():
    """List account's bets"""
    dump(api.list_bets())


@main.command()
def lookup_sports():
    """List Sports"""
    dump(api.lookup_sports())


@main.command()
@click.argument("sport_id", default="1.20.0")
def lookup_eventgroups(sport_id):
    """List event groups in a sport"""
    dump(api.lookup_eventgroups(sport_id))


@main.command()
@click.argument("eventgroup_id", default="1.21.0")
def lookup_events(eventgroup_id):
    """List events in an eventgroup"""
    dump(api.lookup_events(eventgroup_id))


@main.command()
@click.argument("event_id", default="1.22.0")
def lookup_event(event_id):
    """Get an event"""
    dump(api.lookup_event(event_id))


@main.command()
@click.argument("event_id", default="1.22.0")
def lookup_markets(event_id):
    """List markets within an event"""
    dump(api.lookup_markets(event_id))


@main.command()
@click.argument("market_id", default="1.24.0")
def lookup_market(market_id):
    """Get market"""
    dump(api.lookup_market(market_id))


@main.command()
@click.argument("market_id", default="1.24.0")
def lookup_selections(market_id):
    """List all selections from a market"""
    dump(api.lookup_selections(market_id))


@main.command()
@click.argument("selection_id", default="1.25.0")
def lookup_selection(selection_id):
    """Get selection by id"""
    dump(api.lookup_selection(selection_id))


@main.command()
@click.argument("selection_id", default="1.25.0")
def orderbook(selection_id):
    """Show order book of a selection"""
    dump(api.orderbook(selection_id))


@main.command()
@click.argument("selection_id")
@click.argument("back_or_lay", type=click.Choice(["back", "lay"]))
@click.argument("backer_multiplier", type=float)
@click.argument("backer_stake", type=float)
@click.option("--persistent", type=bool)
def placebet(selection_id, back_or_lay, backer_multiplier, backer_stake, persistent):
    """Place a single bet"""
    dump(
        api.place_bet(
            selection_id,
            back_or_lay,
            backer_multiplier,
            "{} BTC".format(backer_stake),
            bool(persistent),
        )
    )


@main.command()
@click.argument("selection_id")
@click.argument("back_or_lay", type=click.Choice(["back", "lay"]))
@click.argument("total_backer_stake", type=float)
@click.argument("mulipliers", type=float, nargs=-1)
@click.option("--persistent", type=bool)
def placebets(persistent, selection_id, back_or_lay, total_backer_stake, mulipliers):
    """Place multiple bets at different odds for a total stake"""
    bets = list()
    for m in mulipliers:
        bets.append(
            (
                selection_id,
                back_or_lay,
                m,
                "{} BTC".format(total_backer_stake / len(mulipliers)),
                bool(persistent),
            )
        )
    dump(api.place_bets(bets))


@main.command()
@click.argument("bet_id")
def cancelbet(bet_id):
    """Cancel a single bet"""
    dump(api.cancel_bet(bet_id))


@main.command()
@click.argument("bet_ids", nargs=-1)
def cancelbets(bet_ids):
    """Cancel many bets"""
    dump(api.cancel_bets(bet_ids))


@main.command()
@click.argument("task_id")
def get_task(task_id):
    """Get the content of a task id"""
    dump(api._get_task(task_id))


@main.command()
@click.argument("sport")
@click.argument("eventgroup")
@click.argument("hometeam")
@click.argument("awayteam")
@click.argument("market")
@click.argument("selection")
def find_market(sport, eventgroup, hometeam, awayteam, market, selection):
    """Find a selection"""
    api.find_selection(
        sport, eventgroup, dict(home=hometeam, away=awayteam), market, selection,
    )


if __name__ == "__main__":
    main()


exbetapi/api.py
# -*- coding: utf-8 -*-
""" API module for Exbet Market Making API
"""
import time
import requests
from .exceptions import (
    JsonDecodingFailedException,
    AlreadyLoggedinException,
    ExecutionError,
    APIError,
)


class ExbetAPI:
    """API Class for Market Making"""

    BASEURL = "https://api.macau.exbet.io/v2/"
    ACCESS_TOKEN = None

    _SLEEPTIME = 0.2
    _BLOCKTIME = 1

    def __init__(self, *args, **kwargs):
        if kwargs.pop("use_everett", None):
            self.BASEURL = "https://mm.api.everett.exbet.io/v2/"
        vars(self).update(**kwargs)

    def _headers(self):
        """Provides header for POST and GET requests"""
        from . import __version__

        headers = dict()
        if self.ACCESS_TOKEN:
            headers["Authorization"] = "Bearer {}".format(self.ACCESS_TOKEN)
        headers["User-Agent"] = "python/exbetapi-" + __version__
        return headers

    def _post(self, endpoint: str, payload: dict) -> dict:
        """Private method to send a POST request with payload to the API"""
        assert isinstance(payload, dict)
        resp = requests.post(
            self.BASEURL + endpoint, json=payload, headers=self._headers()
        )
        try:
            resp.raise_for_status()
        except requests.HTTPError:
            try:
                t = resp.json()
            except Exception:
                t = resp.text
            raise APIError(t)
        try:
            ret = resp.json()
        except Exception:  # pragma: no cover
            raise JsonDecodingFailedException("Decoding API response failed!")
        ret = self._parse_response(ret)
        return ret

    def _get(self, endpoint: str) -> dict:
        """Private method to send a GET from API resource"""
        resp = requests.get(self.BASEURL + endpoint, headers=self._headers())
        try:
            resp.raise_for_status()
        except requests.HTTPError:
            try:
                t = resp.json()
            except Exception:
                t = resp.text
            raise APIError(t)
        try:
            ret = resp.json()
        except Exception:  # pragma: no cover
            raise JsonDecodingFailedException("Decoding API response failed!")
        ret = self._parse_response(ret)
        return ret

    @staticmethod
    def _parse_response(ret):
        """We parse the response for exceptions/errors that the API returns"""
        if "error" in ret:  # pragma: no cover
            raise APIError([x["message"] for x in ret["errors"]])
        return ret

    @staticmethod
    def _require_dict_keys(resp: dict, attributes: list) -> None:
        """Internal test for required dict keys"""
        assert all([x in resp for x in attributes])

    def _get_task(self, task_id: str) -> dict:
        """Obtain a task from the API"""
        return self._post("task", dict(task_id=task_id))

    def _wait_for_task(self, task_id: str) -> dict:  # pragma: no cover
        """Wait for a task to complete"""
        cnt = 0
        while True:
            task = self._get_task(task_id)
            if (
                task.get("ready")
                and task.get("state") == "SUCCESS"
                and task.get("info")
            ):
                task["info"] = eval(task.get("info", "{}"))
                return task
            time.sleep(self._SLEEPTIME)
            cnt += 1
            if cnt > 3 * self._BLOCKTIME / self._SLEEPTIME:
                raise ExecutionError("Inform admins")

    @staticmethod
    def _obtain_task_results(task):  # pragma: no cover
        """A task contains information about the operations performed. Extract
        those
        """
        results = task["info"].get("operation_results")
        if results:
            return [x[1] for x in results]
        return []

    @staticmethod
    def _parse_stake(stake):
        """Stake cane be provided as string and dictionary. Unified here."""
        if isinstance(stake, str) and " " in stake:
            amount, symbol = stake.split(" ")
            return dict(amount=float(amount), symbol=symbol)
        if isinstance(stake, dict) and "amount" in stake and "symbol" in stake:
            return dict(amount=float(stake["amount"]), symbol=stake["symbol"])
        raise ValueError("Invalid stake format")  # pragma: no cover

    #
    # Public methods
    #
    def reset(self):
        """Reset instance"""
        self.ACCESS_TOKEN = None

    def login(self, username: str, password: str) -> None:
        """Obtain and store session for user"""
        if self.ACCESS_TOKEN is not None:
            raise AlreadyLoggedinException("Reset instance with .reset()")
        resp = self._post("login", dict(username=username, password=password))
        self._require_dict_keys(resp, ["access_token", "refresh_token", "username"])
        self.ACCESS_TOKEN = resp["access_token"]

    #
    # General information
    #
    def is_loggedin(self) -> bool:
        """Am i logged in?"""
        return bool(self.ACCESS_TOKEN)

    @property
    def info(self) -> dict:
        """Obtain general information from the API"""
        return self._get("info")

    @property
    def session(self) -> dict:
        """Obtain session information"""
        return self._get("session")

    #
    # Account information
    #
    @property
    def account(self) -> dict:
        """Obtain account details"""
        return self._get("account")

    @property
    def balance(self) -> dict:
        """Obtain account balance from api"""
        balances = self._get("balance").get("balances")
        ret = dict()
        for b in balances:
            ret[b["symbol"]] = b["amount"]
        return ret

    @property
    def roles(self) -> list:
        """Provide roles of the account"""
        return self.account.get("roles")

    #
    # Bets
    #
    def get_bet(self, bet_id: str) -> dict:
        """Obtain details of a placed bet

        :param str bet_id: Bet of the form ``1.26.xxxxxxx``
        """
        return self._post("bet/get", dict(bet_id=bet_id))

    def list_bets(self) -> dict:
        """List account's bets, sorted for matched and unmatched"""
        return self._get("bet/list")

    def cancel_bet(self, bet_id: str, wait=False) -> None:
        """
        :param str bet_id: Bet of the form ``1.26.xxxxxxx``
        :param bool wait: Wait/Block until action has been confirmed
        """
        resp = self._post("bet/cancel", dict(bet_id=bet_id))
        if wait:  # pragma: no cover
            self._wait_for_task(resp.get("task_id"))
        return resp

    def cancel_bets(self, bet_ids: list, wait=False) -> None:
        """
        :param list bet_ids: List of bets of the form ``1.26.xxxxxxx``
        :param bool wait: Wait/Block until action has been confirmed
        """
        resp = self._post(
            "bet/cancel_many",
            dict(bets_to_cancel=[dict(bet_id=bet_id) for bet_id in bet_ids]),
        )
        if wait:  # pragma: no cover
            self._wait_for_task(resp.get("task_id"))
        return resp

    def edit_bet(self, *args, **kwargs):
        """Not yet implemented!"""
        raise NotImplementedError

    def edit_bets(self, *args, **kwargs):
        """Not yet implemented!"""
        raise NotImplementedError

    def place_bet(
        self,
        selection_id: str,
        back_or_lay: str,
        backer_multiplier: float,
        stake: dict,
        persistent=True,
        wait=False,
    ) -> dict:
        """Place a bet

        :param str selection_id: The selection id to place the bet in (1.25.xxxx)
        :param option back_or_lay: Either 'back' or 'lay' the bet
        :param float backer_multiplier: Odds
        :param str stake: Stake as ``amount symbol``
        :param bool persistent: Make the bet persistent over status change from
            upcoming to in_progress
        :param bool wait: Wait for confirmation of the database (if `true`
            returns `bet_id`, else returns `None`).

        Example:::

            bet_id = api.place_bet("1.25.50449", "lay", 1.65, "0.01 BTC", wait=True)

        """
        stake = self._parse_stake(stake)
        resp = self._post(
            "bet/place",
            dict(
                back_or_lay=back_or_lay,
                selection_id=selection_id,
                backer_multiplier=backer_multiplier,
                persistent=persistent,
                backer_stake=stake,
            ),
        )
        if wait:  # pragma: no cover
            return self._obtain_task_results(self._wait_for_task(resp.get("task_id")))[
                0
            ]
        return resp

    def place_bets(self, bets: list, wait=False) -> list:
        """Place multiple bets

        :param list bets: This is a list in the form of `place_bet` arguments.
        :param bool wait: Wait for confirmation of the database (if `true`
                returns `bet_id`, else returns `None`).

        Example:::

            api.place_bets(
                [
                    ("1.25.50449", "lay", 1.65, "0.01 BTC"),
                    ("1.25.50449", "back", 1.75, "0.01 BTC"),
                ],
                wait=True,
            )

        """
        parsed_bets = list()
        for bet in bets:
            b = dict(
                selection_id=bet[0],
                back_or_lay=bet[1],
                backer_multiplier=bet[2],
                backer_stake=self._parse_stake(bet[3]),
                persistent=True,
            )
            if len(bet) > 4:
                b["persistent"] = bet[4]
            parsed_bets.append(b)
        resp = self._post("bet/place_many", dict(place_bets=parsed_bets))
        if wait:  # pragma: no cover
            return self._obtain_task_results(self._wait_for_task(resp.get("task_id")))
        return resp

    #
    # Find
    #
    def find_selection(
        self, sport: str, event_group: str, teams: dict, market: str, selection: str
    ):
        """Find a selection id provided sufficient information

        :param str sport: Sports name (e.g. Basketball)
        :param str event_group: Name of the event group (e.g. NBA)
        :param dict teams: dictionary in the form of ``{"home": "xx", "away": "yy"}``
        :param str market: Market name (e.g. Moneyline)
        :param str selection: The individual selection name (e.g. "xx" for Team xx)

        """
        return self._post(
            "find/selections",
            dict(
                sport=sport,
                event_group=event_group,
                teams=teams,
                market=market,
                selection=selection,
            ),
        )

    #
    # Lookup
    #
    def lookup_selection(self, selection_id: str) -> dict:
        """Provides information about a selection

        :param str selection_id: The selection id (1.25.xxx)
        """
        return self._post("lookup/selection", dict(selection_id=selection_id))

    def lookup_selections(self, market_id: str) -> dict:
        """
        Provides a list of selections in a market

        :param str market_id: Market id (1.24.xxxx)
        """
        return self._post("lookup/selections", dict(market_id=market_id),).get(
            "selections"
        )

    def lookup_market(self, market_id: str) -> dict:
        """Provides information about a market

        :param str market_id: Market  id (1.24.xxxx)
        """
        return self._post("lookup/market", dict(market_id=market_id),)

    def lookup_markets(self, event_id: str) -> list:
        """
        List markets of an event

        :param str event_id: Event id (1.22.xxx)
        """
        return self._post("lookup/markets", dict(event_id=event_id)).get("markets")

    def lookup_event(self, event_id: str) -> dict:
        """Provide information about an event

        :param str event_id: Event id (1.22.xxx)

        """
        return self._post("lookup/event", dict(event_id=event_id))

    def lookup_events(self, eventgroup_id: str) -> list:
        """List events within an event group

        :param str eventgroup_id: Event Group id (1.21.xxx)
        """
        return self._post("lookup/events", dict(eventgroup_id=eventgroup_id)).get(
            "events"
        )

    def lookup_eventgroups(self, sport_id: str) -> list:
        """List event groups with a sport

        :param str sport_id: Sport id (1.20.xxx)
        """
        return self._post("lookup/eventgroups", dict(sport_id=sport_id)).get(
            "eventgroups"
        )

    def lookup_sports(self) -> list:
        """List all sports"""
        return self._post("lookup/sports", dict()).get("sports")

    #
    # Orderbook
    #
    def orderbook(self, selection_id: str) -> dict:
        """Provide (consolidated) order book of a selection

        :param str selection_id: The selection id (1.25.xxx)
        """
        return self._post("lookup/orderbook", dict(selection_id=selection_id))
